src/crypto/CHIPCryptoPAL.h

```diff
@@ -74,17 +74,20 @@ constexpr size_t kEmitDerIntegerWithoutTagOverhead = 1; // 1 sign stuffer
 constexpr size_t kEmitDerIntegerOverhead           = 3; // Tag + Length byte + 1 sign stuffer
 
 /*
- * Worst case is OpenSSL, so let's use its worst case and let static assert tell us if
- * we are wrong, since `typedef SHA_LONG unsigned int` is default.
+ * Size of a static intance of the SHA256 context.
+ *
+ * This must account for the worst case size for all platforms. Currently the
+ * worst case size is the TI hardware accelerated version of mbedtls. This is
+ * 76 words long for the driver context. The previous worst case was OpenSSL
+ * with the following sturuct sizing.
  *   SHA_LONG h[8];
  *   SHA_LONG Nl, Nh;
  *   SHA_LONG data[SHA_LBLOCK]; // SHA_LBLOCK is 16 for SHA256
  *   unsigned int num, md_len;
  *
- * We also have to account for possibly some custom extensions on some targets,
- * especially for mbedTLS, so an extra sizeof(uint64_t) is added to account.
+ * Let the static assert tell us if we are wrong.
  */
-constexpr size_t kMAX_Hash_SHA256_Context_Size = ((sizeof(unsigned int) * (8 + 2 + 16 + 2)) + sizeof(uint64_t));
+constexpr size_t kMAX_Hash_SHA256_Context_Size = (sizeof(unsigned int) * 76);
```

Comment:
The added comment does not fix the fact that we have to allocate this on every platform:

```
The added comment does not fix the fact that we have to allocate this in the `Hash_SHA256_stream` on every platform:

```
struct alignas(size_t) HashSHA256OpaqueContext
{
    uint8_t mOpaque[kMAX_Hash_SHA256_Context_Size];
};
```

Given that this PR appears to increase the worst case from 144 bytes to 304 bytes, I think this PR needs to address that.

Reply:
I agree that this should be address, but I am not quite confident in my understanding of OpenSSL.

Would it be acceptable if I just add a platform `#define` to work around the issue that the Matter crypto code is not actually dependent on the platform code?

Reply:
I would be OK with any solution that only increases the buffer size to 304 bytes on platforms that need this for TI hardware acceleration.

Reply:
The current way this code is setup requires the lowest common denominator for the size of the structure. I agree that this overhead should not be on any other platforms, but it seems like there has been some effort to remove any dependency on platform specific code. So I cannot just `sizeof(sha_ctx)` as I would like, nor can I just define it in my platform configuration header files.

It will take me a while to figure out how to propagate this configuration through the GN build.

Reply:
Then I think you are just the unlucky person who is happening to need a size now that makes the current approach unreasonable.

If other reviewers disagree with me blocking this PR on this issue, please let me know.

Reply:
I agree that this is too large.

I strongly believe that the underlying implementation should shield users from the larger size. I also believe that if concurrent access is done or `GetDigest()` is called, perhaps the driver should go back to software SHA after extracting current digest context. This is what Espressif did to fix concurrent access issues without mutexes or other problems.

Mutex usage is NOT feasible because: the CASE protocol requires long running message transcript over the entire set of messages, for each CASE session establishment. This would prevent other usages of SHA256 from taking place during CASE (but they are required for compressed fabric ID and Destination Node ID computation, as well as for Group Key derivation (though HKDF)), which could deadlock.. Overall, SHA256 must be usable from multiple concurrent contexts transparently.

Reply:
I am not sure concurrency is fixed. I do not see clear handling of concurrency to shield the users.

Reply:
The hardware itself does not preserve state between calls. There's a mutex in the SHA driver to prevent concurrent access of the hardware between contexts. But the object has the intermediate digest stored in a buffer, in each associated context. 

The individual contexts don't lock the mutex for the hardware at the SHA begin and unlock it at the SHA finalize. Instead they lock/unlock around the individual calls to digest bytes and the intermediate context is stored in a buffer in the object. The mutex in the driver ensures that concurrent calls to the hardware won't step on each other, but the calls can be interleaved since the hardware isn't persisting the state. This means that the CASE SHA context can be responsible for digesting multiple fragmented buffers received seconds apart, while another context can be calculating keys with HKDF.